File: kotlin/src/main/kotlin/aethernet/bittorrent/Utp.kt
// SPDX-License-Identifier: MIT

package aethernet.bittorrent

/** A µTP packet type (BEP-29). */
object UtpPacketType {
    const val DATA = 0
    const val FIN = 1
    const val STATE = 2
    const val RESET = 3
    const val SYN = 4
}

/** The µTP protocol version this SDK speaks. */
const val UTP_VERSION = 1

/** The fixed µTP header length. */
const val UTP_HEADER_SIZE = 20

/**
 * A µTP packet (BEP-29, version 1). The 20-byte header is
 * type|version(1) · extension(1) · connection_id(2) · timestamp_us(4) ·
 * timestamp_diff_us(4) · wnd_size(4) · seq_nr(2) · ack_nr(2), all big-endian.
 * The 32-bit fields are held in Longs so the full unsigned range round-trips.
 * The Kotlin port of `go/bittorrent/utp.go`.
 */
class UtpPacket(
    val type: Int,
    val connectionId: Int,
    val timestampMicros: Long,
    val timestampDiff: Long,
    val windowSize: Long,
    val seqNr: Int,
    val ackNr: Int,
    val payload: ByteArray,
) {
    /** Serializes the packet (no extensions). */
    fun toBytes(): ByteArray {
        val buf = ByteArray(UTP_HEADER_SIZE + payload.size)
        buf[0] = ((type shl 4) or UTP_VERSION).toByte()
        buf[1] = 0 // no extensions
        putU16BE(buf, 2, connectionId)
        putU32BE(buf, 4, timestampMicros)
        putU32BE(buf, 8, timestampDiff)
        putU32BE(buf, 12, windowSize)
        putU16BE(buf, 16, seqNr)
        putU16BE(buf, 18, ackNr)
        System.arraycopy(payload, 0, buf, UTP_HEADER_SIZE, payload.size)
        return buf
    }

    companion object {
        /** Parses a µTP packet, walking any extension chain to find the payload. */
        fun parse(data: ByteArray): UtpPacket {
            require(data.size >= UTP_HEADER_SIZE) {
                "µTP packet is ${data.size} bytes, shorter than the $UTP_HEADER_SIZE-byte header"
            }
            val version = data[0].toInt() and 0x0F
            require(version == UTP_VERSION) { "unsupported µTP version $version" }
            val type = (data[0].toInt() and 0xff) ushr 4

            // Walk the extension chain (each: next_ext(1) len(1) data(len)).
            var offset = UTP_HEADER_SIZE
            var nextExt = data[1].toInt() and 0xff
            while (nextExt != 0) {
                require(offset + 2 <= data.size) { "truncated µTP extension header" }
                val thisNext = data[offset].toInt() and 0xff
                val extLen = data[offset + 1].toInt() and 0xff
                offset += 2 + extLen
                require(offset <= data.size) { "truncated µTP extension data" }
                nextExt = thisNext
            }

            return UtpPacket(
                type = type,
                connectionId = u16BE(data, 2),
                timestampMicros = u32BE(data, 4),
                timestampDiff = u32BE(data, 8),
                windowSize = u32BE(data, 12),
                seqNr = u16BE(data, 16),
                ackNr = u16BE(data, 18),
                payload = data.copyOfRange(offset, data.size),
            )
        }
    }
}

File: python/aethernet/bittorrent/utp.py
# SPDX-License-Identifier: MIT
"""µTP packet (BEP-29, version 1) — byte-exact 20-byte header."""
from __future__ import annotations

import struct

DATA = 0
FIN = 1
STATE = 2
RESET = 3
SYN = 4

VERSION = 1
HEADER_SIZE = 20


class UtpPacket:
    def __init__(self, type: int, conn_id: int = 0, timestamp: int = 0, timestamp_diff: int = 0,
                 window: int = 0, seq: int = 0, ack: int = 0, payload: bytes = b""):
        self.type = type
        self.conn_id = conn_id
        self.timestamp = timestamp
        self.timestamp_diff = timestamp_diff
        self.window = window
        self.seq = seq
        self.ack = ack
        self.payload = payload

    def to_bytes(self) -> bytes:
        h = bytearray(HEADER_SIZE)
        h[0] = (self.type << 4) | VERSION
        h[1] = 0  # no extensions
        struct.pack_into(">H", h, 2, self.conn_id)
        struct.pack_into(">I", h, 4, self.timestamp)
        struct.pack_into(">I", h, 8, self.timestamp_diff)
        struct.pack_into(">I", h, 12, self.window)
        struct.pack_into(">H", h, 16, self.seq)
        struct.pack_into(">H", h, 18, self.ack)
        return bytes(h) + self.payload

    @staticmethod
    def parse(data: bytes) -> "UtpPacket":
        if len(data) < HEADER_SIZE:
            raise ValueError(f"µTP packet is {len(data)} bytes, shorter than {HEADER_SIZE}")
        version = data[0] & 0x0F
        if version != VERSION:
            raise ValueError(f"unsupported µTP version {version}")
        packet_type = data[0] >> 4
        offset = HEADER_SIZE
        next_ext = data[1]
        while next_ext != 0:
            if offset + 2 > len(data):
                raise ValueError("truncated µTP extension header")
            this_next = data[offset]
            ext_len = data[offset + 1]
            offset += 2 + ext_len
            if offset > len(data):
                raise ValueError("truncated µTP extension data")
            next_ext = this_next
        return UtpPacket(
            packet_type,
            struct.unpack(">H", data[2:4])[0],
            struct.unpack(">I", data[4:8])[0],
            struct.unpack(">I", data[8:12])[0],
            struct.unpack(">I", data[12:16])[0],
            struct.unpack(">H", data[16:18])[0],
            struct.unpack(">H", data[18:20])[0],
            bytes(data[offset:]),
        )
